File: Python/221. MaximalSquare.py
class Solution(object):
    def maximalSquare(self, matrix):
        # handle 0 and 1 dimensional arrays
        if len(matrix) == 0: return 0
        if len(matrix) == 1:
            for i in range(len(matrix[0])):
                if matrix[0][i] == '1': return 1
            return 0
            
        # input array is not guaranteed to be square. use the smallest side for
        # largest possible square
        max_side = len(matrix)
        if max_side > len(matrix[0]): max_side = len(matrix[0])

        # look for largest possible square to smallest 
        for side in reversed(range(0, max_side)):
            for j in range(len(matrix) - side):
                for k in range(len(matrix[j]) - side):
                    if self.find_square(matrix, j, k, side + 1):
                        return (side + 1)**2
        
        return 0
        
    # check a region in the vector for the existence of a square made of '1'
    def find_square(self, matrix, x, y, side):
        if len(matrix) < 2: return False
        if x + side > len(matrix) or y + side > len(matrix[0]): return False
        
        for i in range(x, x + side):
            for j in range(y, y + side):
                if matrix[i][j] == '0': return False

        return True

File: C++/221. Maximal Square.cpp
/*
Given a 2D binary matrix filled with 0's and 1's, find the largest square containing only 1's and return its area.

Example:

Input: 

1 0 1 0 0
1 0 1 1 1
1 1 1 1 1
1 0 0 1 0

Output: 4
*/

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int r = matrix.size();
        if(r==0)
            return 0;
        int c = matrix[0].size();
        int dp[r+1][c+1];
        for(int i=0;i<=r;i++){
            for(int j=0;j<=c;j++){
                    dp[i][j]=0;
            }
        }
        int ans=0;
        for(int i=1;i<=r;i++){
            for(int j=1;j<=c ;j++)
                if(matrix[i-1][j-1]=='1'){
                    dp[i][j] = 1 + min(dp[i-1][j-1],min(dp[i-1][j],dp[i][j-1]));
                    if(ans<dp[i][j])
                        ans=dp[i][j];
                }
        }
        return ans*ans;
        
    }
};
